### PR TITLE
Add ClaimAuthorizeService

### PR DESCRIPTION
**Pull request by arnaucube**
Add ClaimAuthorizeService, that is a claim to authorize a Service for the identity that performs the claim.

```
e_3: [ 64  bits] claim type
     [ 32  bits] version
     [ 157 bits] 0
e_2: [ 248 bits] Hash_k(addr)
     [ 5   bits] 0
e_1: [ 248 bits] Hash_k(pubK)
     [ 5   bits] 0
e_0: [ 248 bits] Hash_k(domain)
     [ 5   bits] 0
```
Where `hash` is the Keccak(input_string)[-248/8:]
```
[0 | Hash_k(domain)][0 | Hash_k(pubK)][0 | Hash_k(addr)][0 | ver | type]
      e0                    e1              e2                  e3
```

This claim is used to authorize services to act for the identity that performs the claim, for example, with this claim type, an identity can authorize a Relay, a NotificationsServer, etc to act as their authorized services.
Also, when the identity is discovered through the discovery-protocol, will be provided the proofs of this claim, so a requester can verifiy that a service is authorized by the identity in the identity's merkletree.

### CODE REVIEW

**Comment by ed255** [comment]
Shouldn't the claim include the actions that the entity allows the service to perform?
- Receive notifications on behalf of the user
- Add entity root keys through SetRoot Claim (Relay)
- ...

Also, what is "addr", "pubK", and "domain"?

**Comment by arnaucube** [comment]
**Update**: Added `ServiceType` to the ClaimAuthorizeService ( [0ad04429ffd0ae1e84c1139b2657b9c5fbe0e05a](https://github.com/iden3/go-iden3/pull/114/commits/0ad04429ffd0ae1e84c1139b2657b9c5fbe0e05a) ), now the Claim data structure is:
```
e_3: [ 64  bits] claim type
     [ 32  bits] version
     [ 64  bits] service type
     [ 93 bits] 0
e_2: [ 248 bits] Hash_k(addr)
     [ 5   bits] 0
e_1: [ 248 bits] Hash_k(pubK)
     [ 5   bits] 0
e_0: [ 248 bits] Hash_k(url)
     [ 5   bits] 0
```
```
0 <---> N bytes
[0 | Hash_k(url)][0 | Hash_k(pubK)][0 | Hash_k(addr)][0 | service type | ver | type]
      e0                    e1              e2                       e3
```

File: core/claim.go [hunk 1]
```diff
@@ -53,6 +53,13 @@ func GetClaimTypeVersionFromData(d *merkletree.Data) (c ClaimType, v uint32) {
 	return c, v
 }
 
+// HashString takes the first 31 bytes of a hash applied to name.
```
Comment (krlosMata) [inline core/claim.go]: I would change the comment:
`...hash applied to a string` instead of `...hash applied to name`

Comment (arnaucube) [reply core/claim.go]: done :)

File: core/claim.go [hunk 2]
```diff
@@ -466,6 +468,83 @@ func (c *ClaimLinkObjectIdentity) Type() ClaimType {
...
+	// Version is the claim version.
+	Version uint32
+	// ServiceType is the type of the authorized service
+	ServiceType *ServiceType
+	// ServiceAddr is the hash of the addr
+	ServiceAddr [248 / 8]byte
+	// ServicePubK is the hash of the pubK
+	ServicePubK [248 / 8]byte
+	// ServiceDomain is the hash of the domain
+	ServiceDomain [248 / 8]byte
```
Comment (ed255) [inline core/claim.go]: Shouldn't this be ServiceUrl?

File: core/claim.go [hunk 3]
```diff
@@ -466,6 +468,83 @@ func (c *ClaimLinkObjectIdentity) Type() ClaimType {
...
+}
+
+// NewClaimAuthorizeService returns a ClaimAuthorizeService with the provided data.
+func NewClaimAuthorizeService(serviceType *ServiceType, serviceAddr, servicePubK, serviceDomain string) *ClaimAuthorizeService {
+	return &ClaimAuthorizeService{
+		Version:       0,
+		ServiceType:   serviceType,
+		ServiceAddr:   HashString(serviceAddr),
+		ServicePubK:   HashString(servicePubK),
+		ServiceDomain: HashString(serviceDomain),
```
Comment (ed255) [inline core/claim.go]: ServiceUrl

File: core/claim.go [hunk 4]
```diff
@@ -466,6 +468,83 @@ func (c *ClaimLinkObjectIdentity) Type() ClaimType {
...
+// NewClaimAuthorizeServiceFromEntry deserializes a ClaimAuthorizeService from an Entry.
+func NewClaimAuthorizeServiceFromEntry(e *merkletree.Entry) *ClaimAuthorizeService {
+	c := &ClaimAuthorizeService{}
+	_, c.Version = getClaimTypeVersion(e)
+	var serviceType [64 / 8]byte
+	copyFromElemBytes(serviceType[:], ClaimTypeVersionLen, &e.Data[3])
+	c.ServiceType = NewServiceType(binary.BigEndian.Uint64(serviceType[:]))
+	copyFromElemBytes(c.ServiceAddr[:], 0, &e.Data[2])
+	copyFromElemBytes(c.ServicePubK[:], 0, &e.Data[1])
+	copyFromElemBytes(c.ServiceDomain[:], 0, &e.Data[0])
```
Comment (ed255) [inline core/claim.go]: ServiceUrl

File: core/claim.go [hunk 5]
```diff
@@ -466,6 +468,83 @@ func (c *ClaimLinkObjectIdentity) Type() ClaimType {
...
+}
+
+// Entry serializes the claim into an Entry.
+func (c *ClaimAuthorizeService) Entry() *merkletree.Entry {
+	e := &merkletree.Entry{}
+	setClaimTypeVersion(e, c.Type(), c.Version)
+	copyToElemBytes(&e.Data[3], ClaimTypeVersionLen, c.ServiceType[:])
+	copyToElemBytes(&e.Data[2], 0, c.ServiceAddr[:])
+	copyToElemBytes(&e.Data[1], 0, c.ServicePubK[:])
+	copyToElemBytes(&e.Data[0], 0, c.ServiceDomain[:])
```
Comment (ed255) [inline core/claim.go]: ServiceUrl

File: core/claim.go (hunk 2 above)
Comment (arnaucube) [reply core/claim.go]: right

**Pull request merged**